Vmi fix addressspace

Wenzel opened the pull request:
This fixes the following issues:
* start address must be at 0
* address space must be `Volatile`
* add support to specify the hypervisor (xen or kvm)
* handle cases where read fails

One thing we could discuss is that the user could specify if he wants the address space to be `Volatile` or not.
We could pause the VM and resume it to get a consistent memory access.

Example:
~~~Python
def __init__()
    self.vmi = Libvmi()
    self.vmi.pause()


def close()
    self.vmi.destroy()
    # destroy will resume the vm automatically
~~~

scudette commented on rekall-core/rekall/plugins/addrspaces/vmi.py:
```diff
@@ -24,21 +36,30 @@ def __init__(self, base=None, filename=None, session=None, **kwargs):
+        try:
+            buffer, bytes_read = self.vmi.read_pa(addr, size)
+        except LibvmiError:
+            buffer = bytes(size)
```
Please use the zeroer to make this more efficient.

https://github.com/google/rekall/blob/master/rekall-core/rekall/addrspace.py#L55

scudette commented on rekall-core/rekall/plugins/addrspaces/vmi.py:
```diff
@@ -24,21 +36,30 @@ def __init__(self, base=None, filename=None, session=None, **kwargs):
-        if bytes_read != size:
-            raise RuntimeError('Error while reading physical memory at {}'.format(hex(addr)))
+        try:
+            buffer, bytes_read = self.vmi.read_pa(addr, size)
```
It is up to you but if you already know the read will fall outside the allowed range you might want to avoid calling into libvmi because it cause it to log errors anyway.

I am not sure how the underlying library runs the reads but maybe its worth breaking the reads up into parts which fall within allowable ranges and those which should be zero filled. The easiest way to do this is to overlay a run based address space on top of the raw libvmi address space and then program the run based address space with the allowed runs. Can you get the list of valid runs from libvmi or is it capable of reading everywhere and null padding by itself?

Wenzel replied:
I'm calling @tklengyel here since he knows better libvmi internals =)

tklengyel replied:
LibVMI doesn't pad reads to match what the requested size was. If you are reading such that it spans multiple pages, and one of those pages is unmappable, it will error out and return to you as much as it was able to read and tell you what that was in `bytes_read`, see https://github.com/libvmi/libvmi/blob/master/libvmi/read.c#L118. If you need a padded read, you can implement a wrapper around this to do so.

Wenzel replied:
I could implement this padded read directly in the Libvmi python bindings instead of copying the definition in both Rekall and Volatility as.

scudette replied:
Is there a way to get the mapped pages from libvmi in advance? Otherwise you could just break up the reads into separate page size reads and issue a single call per page - then if the read fails you just pad with a single page of zeros for that read and keep trying the rest of the range.

This would slow down large scanning plugins (which typically read 10mb at a time) but probably wont make too much difference for plugins which read a few bytes at a time.

Wenzel replied:
>It is up to you but if you already know the read will fall outside the allowed range you might want to avoid calling into libvmi because it cause it to log errors anyway.
I would prefer that also, because libvmi will generate errors on stderr:
~~~
VMI_ERROR: --requesting PA [0xff001000] beyond max physical address [0xff000000]
VMI_ERROR: 	paddr: ff000000, length 1000, vmi->max_physical_address ff000000
VMI_ERROR: create_new_entry failed
~~~

Why does Rekall tries to read outside of the given mappings that we have defined in the AddressSpace with `get_mappings()/is_valid_address()/get_available_addresses` ?
When these methods are used inside Rekall ?

> Is there a way to get the mapped pages from libvmi in advance?
You can get the mapped pages of a specific process:
https://github.com/libvmi/libvmi/blob/master/libvmi/libvmi_extra.h#L46
Not sure that's what you were asking by mapped pages ?

scudette replied:
I was more referring to the physical memory pages which are backed by actual RAM. Not all of physical memory is readable and there are only some valid mapped regions which are backed by memory (e.g. the output of /proc/iomem - the areas marked as System RAM).

Rekall should not be reading outside these regions normally but this appears to be what is happening here. Can you verify that get_mappings() returns the correct regions - the error message above seems to indicate that the maximum physical address should be 0xff000000. 

I guess this is a good point - we should answer why Rekall is trying to access outside these regions in the first place. I suspect that there is some inconsistency between the return of get_mappings() and the allowed physical address regions.

Wenzel replied:
>we should answer why Rekall is trying to access outside these regions in the first place. I suspect that there is some inconsistency between the return of get_mappings() and the allowed physical address regions.

When rekall calls the read method, is there a filter (like `is_valid_address` beforehand to check if the requested address is in the range that we defined ?

If so, does it check if `addr` is in the range, or `addr` + `count` ?

Any idea how i can start to get some information ?

Wenzel commented on rekall-core/rekall/plugins/addrspaces/vmi.py:
```diff
@@ -24,21 +36,30 @@ def __init__(self, base=None, filename=None, session=None, **kwargs):
-                       "The domain must be passed with the URL prefix {}".format(URL_PREFIX))
-        domain = url[len(URL_PREFIX):]
-        self.as_assert(domain, "domain name missing after {}".format(URL_PREFIX))
+        pattern = '{}((?P<hypervisor>xen|kvm)/)?(?P<domain>.*)'.format(URL_PREFIX)
```
this regex is not in a loop.
is it really a performance issue ? :thinking:

Wenzel commented on rekall-core/rekall/plugins/addrspaces/vmi.py:
```diff
@@ -24,21 +36,30 @@ def __init__(self, base=None, filename=None, session=None, **kwargs):
-                       "The domain must be passed with the URL prefix {}".format(URL_PREFIX))
-        domain = url[len(URL_PREFIX):]
-        self.as_assert(domain, "domain name missing after {}".format(URL_PREFIX))
+        pattern = '{}((?P<hypervisor>xen|kvm)/)?(?P<domain>.*)'.format(URL_PREFIX)
```
I didn't know about the possibilities of `urlparse`.
Now we can add parameters like the `volatile` one to the URL :)

Wenzel commented:
@scudette while we are trying to find how to fix the `read` issues, i updated the PR to use `urlparse` instead of a regex.

Also i added support to parse `volatile` as a query parameter, allowing the user to specify if he wants to run on a consistent memory, therefore pausing the VM.

One thing however, is that i found the method `close()` has never been called.... :thinking: 
So i cannot resume the VM properly, nor destroy the vmi_instance.

When `close()` is supposed to be called on a AddressSpace ?

Wenzel commented:
@scudette little update to clarify the state of this PR.

TODO:
- [ ] Find out why `close()` is not called when the AddressSpace is destroyed
- [ ] Review the `volatile` parameter, if the domain is pause, how should we cache the results
- [ ] Investigate the `read` issue, why the method is called with an address outside of the mappings that we defined.

Wenzel commented:
I added a `print('#### close ####')` into the `close()` method of `FDAddressSpace`,
https://github.com/google/rekall/blob/master/rekall-core/rekall/plugins/addrspaces/standard.py#L98
And a `print('#### opening ####)` in the `__init__` method:

~~~
$ rekall -f nitro_win7x64.raw pslist
#### opening ####
#### opening ####
2018-03-26 15:25:29,889:WARNING:rekall.1:Unable to determine file size, assuming file is volatile.
#### opening ####
#### opening ####
#### opening ####
  _EPROCESS            name          pid   ppid  thread_count handle_count session_id wow64    process_create_time       process_exit_time    
-------------- -------------------- ----- ------ ------------ ------------ ---------- ------ ------------------------ ------------------------
#### opening ####
0xfa8001108890 System                   4      0           76          467          - False  2018-03-26 22:12:53Z     -                       
#### opening ####
0xfa8002144040 smss.exe               248      4            2           29          - False  2018-03-26 22:12:53Z     -                       
#### opening ####
#### opening ####
0xfa80011c5b30 svchost.exe            300    464           12          335          0 False  2018-03-26 22:14:56Z     -                       
#### opening ####
#### opening ####
0xfa8002159060 csrss.exe              324    308            8          277          0 False  2018-03-26 22:12:54Z     -                       
#### opening ####
#### opening ####
0xfa80021c52d0 svchost.exe            344    464           15          394          0 False  2018-03-26 22:12:55Z     -                       
#### opening ####
#### opening ####
0xfa8002b7c060 wininit.exe            372    308            3           74          0 False  2018-03-26 22:12:54Z     -                       
#### opening ####
#### opening ####
~~~

But there is not `#### closing  ####`, so it's not just me, there is a problem on how rekall is calling this method to destroy the address spaces.

scudette commented:
Python does not have destructors so we dont call the close on destruction but when the address space is used as a context manager. Usually we dont bother to close the address space when Rekall is used in this way but sometimes it is important - for example when using the live mode we need to close the address space in order to be able to unload the driver - so we do it deliberately:
https://github.com/google/rekall/blob/master/rekall-core/rekall/plugins/tools/live_windows.py#L212

There are other cases when it is important to close the address space mostly around running in the Agent and then we close the address space too. But it is not normally closed automatically.

The volatile parameter only affects the type of cache that is used - volatile just selects the timed cache:
https://github.com/google/rekall/blob/master/rekall-core/rekall/cache.py#L122

If you want to get very sophisticated you can either implement your own cache based on the timed cache or just tweak the timeout parameter in the cache. What do you want to happen then the domain is paused? Can you detect when it is unpaused at runtime?

Wenzel commented:
> There are other cases when it is important to close the address space mostly around running in the Agent and then we close the address space too. But it is not normally closed automatically.

The problem is that i allocated space with my `vmi_instance` and i need to destroy it properly.

What about using `atexit.register()` ?
https://docs.python.org/3/library/atexit.html#atexit.register

> Python does not have destructors

What about implementing `__del__()` ? 

> Can you detect when it is unpaused at runtime?

No, i can't do that. The domain is expected to stay paused, and the AddressSpace will unpause it when it's done.

> What do you want to happen then the domain is paused? 

I just want to work on a consistent physical memory. This is when `volatile = False`, the domain is paused, and the cache doesn't have to be time based, since the memory content stay the same over the course of the session.

scudette commented:
Are you planning to run Rekall itself or use Rekall as a library?

You can see here that we call session.Flush() before we exit:
https://github.com/google/rekall/blob/master/rekall-core/rekall/rekal.py#L111

and you can register a flush hook in the session object:
https://github.com/google/rekall/blob/master/rekall-core/rekall/session.py#L1110

When running in your own script you can close the address space by itself or simply call session.Flush() when done.

Wenzel commented:
> Are you planning to run Rekall itself or use Rekall as a library?

Both of them !

In the latest commit i registered a hook to call `close()`.
I tested and the hook is called.


Now i have a new issue with the `volatile` parameter. When i disable it, it seems that Rekall is trying to read from the address space even after the `session.Flush()`:
Here, you can see that `### close ###` has been called, the vmi instance has been destroyed, and
 `self.vmi` has been set to None:
~~~
$ rekall -f 'vmi:///nitro_win7x64?volatile=0' pslist
xc: error: Could not obtain handle on privileged command interface (2 = No such file or directory): Internal error
VMI_ERROR: Failed to open libxc interface.
LibVMI Version 0.11.0
LibVMI Driver Mode 1
--completed driver init.
--got id from name (nitro_win7x64 --> 4)
**set image_type = nitro_win7x64
--libvirt version 1003001
--qmp: virsh -c qemu:///system qemu-monitor-command nitro_win7x64 '{"execute": "pmemaccess", "arguments": {"path": "/tmp/vmiBtifqf"}}'
--kvm: using custom patch for fast memory access
**set allocated_ram_size = 59700000, max_physical_address = 0x59700000
  _EPROCESS            name          pid   ppid  thread_count handle_count session_id wow64    process_create_time       process_exit_time    
-------------- -------------------- ----- ------ ------------ ------------ ---------- ------ ------------------------ ------------------------
0xfa8001108890 System                   4      0           77          466          - False  2018-04-05 02:20:47Z     -                       
0xfa8002290b30 smss.exe               248      4            2           29          - False  2018-04-05 02:20:47Z     -                       
0xfa8002a4d060 csrss.exe              324    316            9          300          0 False  2018-04-05 02:20:49Z     -                       
0xfa8002d5eb30 svchost.exe            364    468           18          407          0 False  2018-04-05 02:20:50Z     -                       
0xfa800116b060 wininit.exe            372    316            3           74          0 False  2018-04-05 02:20:49Z     -                       
0xfa800116d460 csrss.exe              384    364            7          127          1 False  2018-04-05 02:20:49Z     -                       
0xfa8002afa4f0 winlogon.exe           424    364            3          109          1 False  2018-04-05 02:20:49Z     -                       
0xfa8002269b30 services.exe           468    372            5          177          0 False  2018-04-05 02:20:49Z     -                       
0xfa8002bec910 lsass.exe              476    372            6          514          0 False  2018-04-05 02:20:49Z     -                       
0xfa8002bf1b30 lsm.exe                484    372            9          135          0 False  2018-04-05 02:20:49Z     -                       
0xfa8002c3e9e0 svchost.exe            588    468            9          333          0 False  2018-04-05 02:20:49Z     -                       
0xfa8002c41b30 svchost.exe            652    468            6          205          0 False  2018-04-05 02:20:49Z     -                       
0xfa8002c83060 svchost.exe            704    468           19          440          0 False  2018-04-05 02:20:49Z     -                       
0xfa8002cac4e0 svchost.exe            796    468           17          409          0 False  2018-04-05 02:20:50Z     -                       
0xfa80026f95c0 svchost.exe            864    468           22          820          0 False  2018-04-05 02:20:50Z     -                       
0xfa8002d30b30 svchost.exe           1016    468           12          299          0 False  2018-04-05 02:20:50Z     -                       
0xfa80029c1890 spoolsv.exe           1040    468           12          263          0 False  2018-04-05 02:20:50Z     -                       
0xfa8002de0b30 svchost.exe           1080    468           19          308          0 False  2018-04-05 02:20:50Z     -                       
0xfa8002e3bb30 taskhost.exe          1164    468            7          143          1 False  2018-04-05 02:20:50Z     -                       
0xfa8002e8a060 svchost.exe           1368    468            8          154          0 False  2018-04-05 02:20:51Z     -                       
0xfa8002ec76b0 sppsvc.exe            1428    468            4          142          0 False  2018-04-05 02:20:51Z     -                       
0xfa8002e92360 wlms.exe              1512    468            4           43          0 False  2018-04-05 02:20:51Z     -                       
0xfa8002fed5a0 dwm.exe               1764    796            3           68          1 False  2018-04-05 02:22:50Z     -                       
0xfa8002fb2420 svchost.exe           1776    468            5           89          0 False  2018-04-05 02:20:51Z     -                       
0xfa8002ddc590 explorer.exe          1860   1552           18          608          1 False  2018-04-05 02:22:50Z     -                       
0xfa8001318060 svchost.exe           1900    468           12          340          0 False  2018-04-05 02:22:51Z     -                       
0xfa8001344b30 WmiPrvSE.exe          2016    588            7          108          0 False  2018-04-05 02:41:34Z     -                       
### close ###                                                
Traceback (most recent call last):
  File "/home/tarrma/fsecure/rekall/rekall-core/rekall/plugins/addrspaces/intel.py", line 311, in vtop
    return self._tlb.Get(vaddr)
  File "/home/tarrma/fsecure/rekall/rekall-core/rekall/addrspace.py", line 76, in Get
    result = self.page_cache.Get(vaddr & self.PAGE_MASK)
  File "/home/tarrma/fsecure/rekall/rekall-lib/rekall_lib/utils.py", line 147, in NewFunction
    return f(self, *args, **kw)
  File "/home/tarrma/fsecure/rekall/rekall-lib/rekall_lib/utils.py", line 336, in Get
    raise KeyError(key)
KeyError: 272678927695872

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/tarrma/fsecure/rekall/rekall-core/rekall/plugins/addrspaces/intel.py", line 581, in read_pte
    return self._cache.Get(addr)
  File "/home/tarrma/fsecure/rekall/rekall-lib/rekall_lib/utils.py", line 147, in NewFunction
    return f(self, *args, **kw)
  File "/home/tarrma/fsecure/rekall/rekall-lib/rekall_lib/utils.py", line 336, in Get
    raise KeyError(key)
KeyError: 1671328

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/tarrma/tmp/venv-rekall/bin/rekall", line 11, in <module>
    load_entry_point('rekall-core', 'console_scripts', 'rekall')()
  File "/home/tarrma/fsecure/rekall/rekall-core/rekall/rekal.py", line 111, in main
    user_session.Flush()
  File "/home/tarrma/fsecure/rekall/rekall-core/rekall/session.py", line 1136, in Flush
    hook(*args)
  File "/home/tarrma/fsecure/rekall/rekall-core/rekall/cache.py", line 280, in Flush
    self.SetFingerprint(self.session.GetParameter("image_fingerprint"))
  File "/home/tarrma/fsecure/rekall/rekall-core/rekall/session.py", line 768, in GetParameter
    result = self._RunParameterHook(item)
  File "/home/tarrma/fsecure/rekall/rekall-core/rekall/session.py", line 807, in _RunParameterHook
    result = hook.calculate()
  File "/home/tarrma/fsecure/rekall/rekall-core/rekall/plugins/windows/misc.py", line 352, in calculate
    result.append((address_space.vtop(label.obj_offset), label.v()))
  File "/home/tarrma/fsecure/rekall/rekall-core/rekall/plugins/addrspaces/intel.py", line 316, in vtop
    aligned_vaddr, PhysicalAddressDescriptorCollector(self.session))
  File "/home/tarrma/fsecure/rekall/rekall-core/rekall/plugins/addrspaces/amd64.py", line 119, in describe_vtop
    self._describe_pde(collection, pde_addr, vaddr)
  File "/home/tarrma/fsecure/rekall/rekall-core/rekall/plugins/windows/pagefile.py", line 481, in _describe_pde
    pde_value = self.read_pte(pde_addr)
  File "/home/tarrma/fsecure/rekall/rekall-core/rekall/plugins/addrspaces/intel.py", line 583, in read_pte
    string = self.base.read(addr, 8)
  File "/home/tarrma/fsecure/rekall/rekall-core/rekall/plugins/addrspaces/vmi.py", line 68, in read
    return self.vmi.read_pa_padded(addr, size)
  File "/home/tarrma/tmp/venv-rekall/lib/python3.5/site-packages/libvmi-3.0-py3.5-linux-x86_64.egg/libvmi/libvmi.py", line 398, in read_pa_padded
    status = lib.vmi_read_pa(self.vmi, paddr, count, buffer, bytes_read)
TypeError: initializer for ctype 'struct vmi_instance *' must be a cdata pointer, not NoneType
~~~

But Rekall is trying to read from the address space:
~~~
 status = lib.vmi_read_pa(self.vmi, paddr, count, buffer, bytes_read)
~~~

And it fails, ofc.

Any idea why ?

Thanks !

Wenzel commented:
About the `read` issue, i added this code to test is the `addr` is valid before performing the `read` operation:

~~~Python
    def read(self, addr, size):
        if not self.is_valid_address(addr):
            print('Invalid addr {}'.format(hex(addr)))
        buffer, _ = self.vmi.read_pa(addr, size, padding=True)
~~~

results:
~~~
$ sudo ./venv-rekall/bin/rekall -f vmi://xen/xenwin7 pstree
LibVMI Version 0.11.0
LibVMI Driver Mode 0
--completed driver init.
--got id from name (xenwin7 --> 3)
**set image_type = xenwin7
**set allocated_ram_size = 7d04b000, max_physical_address = 0xff000000
Invalid addr 0x9000000330f7800 (0xa00000)        
VMI_ERROR: --requesting PA [0x9000000331000] beyond max physical address [0xff000000]
VMI_ERROR: 	paddr: 9000000330000, length 1000, vmi->max_physical_address ff000000
VMI_ERROR: create_new_entry failed
Invalid addr 0x9000000330330
VMI_ERROR: --requesting PA [0x9000000331000] beyond max physical address [0xff000000]
VMI_ERROR: 	paddr: 9000000330000, length 1000, vmi->max_physical_address ff000000
VMI_ERROR: create_new_entry failed
_EPROCESS                                 ppid  thd_count hnd_count       create_time       
---------------------------------------- ------ --------- --------- ------------------------
  0xfa8002a619d0 csrss.exe (320)            312         8       302 2018-04-05 03:04:25Z    
  0xfa8001817060 wininit.exe (368)          312         3        74 2018-04-05 03:04:25Z    
. 0xfa80035c55f0 services.exe (464)         368         9       189 2018-04-05 03:04:25Z    
.. 0xfa8002982330 svchost.exe (588)         464        12       350 2018-04-05 03:04:26Z    
.. 0xfa8002d24350 svchost.exe (604)         464        19       509 2018-04-05 03:04:26Z    
.. 0xfa80028d6250 svchost.exe (648)         464         7       220 2018-04-05 03:04:26Z    
.. 0xfa8002bcc890 svchost.exe (700)         464        19       403 2018-04-05 03:04:26Z    
.. 0xfa8002c12060 svchost.exe (792)         464        19       408 2018-04-05 03:04:26Z    
... 0xfa8002882060 dwm.exe (768)            792         5        73 2018-04-05 03:06:26Z    
.. 0xfa8002c82b30 svchost.exe (864)         464        34       899 2018-04-05 03:04:26Z    
.. 0xfa8002ce5820 svchost.exe (984)         464        14       272 2018-04-05 03:04:26Z    
.. 0xfa8002875060 spoolsv.exe (1108)        464        14       274 2018-04-05 03:04:27Z    
.. 0xfa8002dc06e0 taskhost.exe (1140)       464         8       145 2018-04-05 03:04:27Z    
.. 0xfa8002dd99b0 svchost.exe (1164)        464        20       311 2018-04-05 03:04:27Z    
.. 0xfa8002e245b0 svchost.exe (1276)        464         8       154 2018-04-05 03:04:27Z    
.. 0xfa8002b41a30 sppsvc.exe (1392)         464         5       144 2018-04-05 03:04:27Z    
.. 0xfa8002ef8800 wlms.exe (1456)           464         4        43 2018-04-05 03:04:27Z    
.. 0xfa8002f8cb30 svchost.exe (1660)        464         5        89 2018-04-05 03:04:28Z    
.. 0xfa800187f9e0 svchost.exe (1908)        464        13       346 2018-04-05 03:06:27Z    
. 0xfa8002af9b30 lsass.exe (472)            368         7       572 2018-04-05 03:04:26Z    
. 0xfa8002b05b30 lsm.exe (480)              368        11       144 2018-04-05 03:04:26Z    
  0xfa8001782890 System (4)                   0        77       450 2018-04-05 03:04:25Z    
. 0xfa8002883710 smss.exe (248)               4         2        29 2018-04-05 03:04:25Z    
  0xfa8001818730 csrss.exe (380)            360         7       197 2018-04-05 03:04:25Z    
  0xfa8002ad94f0 winlogon.exe (420)         360         6       113 2018-04-05 03:04:25Z    
. 0xfa8002f87a90 userinit.exe (772)         420         0         - 2018-04-05 03:06:26Z    
.. 0xfa8002b88340 explorer.exe (1388)       772        30       701 2018-04-05 03:06:26Z    
... 0xfa8002e5ab30 iexplore.exe (1500)     1388        22       606 2018-04-05 03:06:36Z    
.... 0xfa8001a0a1c0 iexplore.exe (456)     1500        11       261 2018-04-05 03:06:37Z    
.... 0xfa80019ff060 iexplore.exe (1084)    1500        24       591 2018-04-05 03:06:39Z 
~~~

How can we debug this and understand why he is trying to read outside of the mappings ?

Wenzel commented:
@scudette  i think we could already merge this as it brings good improvments, and some scripts in the libvmi repository depends on these fixes.

We can continue with the `volatile` and the `read` bugs in another issue maybe ?

scudette commented:
Yes I agree. I think there may be a small bug in some of the reading code. 0x9000000331000 looks like its is probably trying to read 0x000000331000 but it might have a high bit set somewhere. In reality the address should be converted to a pointer which is only 48 bits long and so the top bits should be cleared. If you have time I would love to see a back trace when that read occurs (insert a pdb.set_trace() at the print site and type "bt" to get a backtrace). I would bet this is probably a missing call to Pointer.integer_to_address().

Lets merge this one though - sorry for the delay in getting this in.
